src/main/kotlin/org/rust/lang/core/types/infer/Fold.kt

```diff
@@ -89,14 +99,27 @@ fun <T> TypeFoldable<T>.foldTyProjectionWith(folder: (TyProjection) -> Ty): T =
  */
 fun <T> TypeFoldable<T>.substitute(subst: Substitution): T =
     foldWith(object : TypeFolder {
-        override fun invoke(ty: Ty): Ty =
-            subst[ty] ?: if (ty.hasTyTypeParameters) ty.superFoldWith(this) else ty
+        override fun foldTy(type: Ty): Ty {
+            val typeWithLifetime = (type as? TyReference)?.copy(lifetime = foldRe(type.lifetime)) ?: type
+            val value = (typeWithLifetime as? TyTypeParameter)?.let { subst[it] }
+            if (value != null) return value
```

Comment:
Can be simplified to `(typeWithLifetime as? TyTypeParameter)?.let { return subst[it] }`

Reply:
No, because `subst[it]` has type `Ty?`.

Reply:
:thinking: 
`(typeWithLifetime as? TyTypeParameter)?.let { subst[it] }?.let { return it }`
?